cloudinit/config/schema.py

```diff
@@ -390,45 +408,88 @@ def _get_schema_examples(schema, prefix=''):
     return rst_content
 
 
-def get_schema_doc(schema):
-    """Return reStructured text rendering the provided jsonschema.
+def get_meta_doc(meta: MetaSchema, schema: dict) -> str:
+    """Return reStructured text rendering the provided metadata.
 
-    @param schema: Dict of jsonschema to render.
-    @raise KeyError: If schema lacks an expected key.
+    @param meta: Dict of metadata to render.
+    @raise KeyError: If metadata lacks an expected key.
     """
-    schema_copy = deepcopy(schema)
-    schema_copy['property_doc'] = _get_property_doc(schema)
-    schema_copy['examples'] = _get_schema_examples(schema)
-    schema_copy['distros'] = ', '.join(schema['distros'])
+
+    # Some modules set this to falsy value, exit gracefully
+    if not meta or not schema:
+        return ''
+    keys = set(meta.keys())
+    expected = set({
+        'id',
+        'title',
+        'examples',
+        'frequency',
+        'distros',
+        'description',
+        'name'})
+    if keys != expected:
+        raise KeyError(
+            'Missing module metadata key(s) {}'.format(expected - keys))
+
+    # cast away type annotation
+    meta_copy = dict(deepcopy(meta))
+    meta_copy['property_doc'] = _get_property_doc(schema)
+    meta_copy['examples'] = _get_examples(meta)
+    meta_copy['distros'] = ', '.join(meta['distros'])
     # Need an underbar of the same length as the name
-    schema_copy['title_underbar'] = re.sub(r'.', '-', schema['name'])
-    return SCHEMA_DOC_TMPL.format(**schema_copy)
+    meta_copy['title_underbar'] = re.sub(r'.', '-', meta['name'])
+    template = SCHEMA_DOC_TMPL.format(**meta_copy)
+    return template
+
+
+def get_modules() -> dict:
+    configs_dir = os.path.dirname(os.path.abspath(__file__))
+    return find_modules(configs_dir)
+
 
+def load_doc(module: list) -> str:
+    '''Load module docstrings
 
-FULL_SCHEMA = None
+    Docstrings are generated on module load. Reduce, reuse, recycle.
+    '''
+    docs = ''
+    for (_, mod_name) in get_modules().items():
+        if 'all' in module or mod_name in module:
+            (mod_locs, _) = importer.find_module(
+                mod_name, ['cloudinit.config'], ['schema'])
+            if mod_locs:
+                mod = importer.import_module(mod_locs[0])
+                docs += mod.__doc__ or ""
+    return docs
 
 
-def get_schema():
+def get_schema() -> dict:
     """Return jsonschema coalesced from all cc_* cloud-config module."""
-    global FULL_SCHEMA
```

Comment:
This global was previously a reimplementation of features provided by standard library functionality (`lru_cache`). The new implementation uses `load_doc()`, which references the docstrings that are already computed on module load. This means that, like the previous implementation, docstrings are only generated once.

Reply:
Docstrings are only generated once, but get_schema is still recalculating what full_schema is each time it is called. If this function was called multiple times throughout cloud-init bringup, we'd probably want to put an `@lru_cache` on  this I think, but since we only have one call site for get_schema() `validate_cloudconfig_file` and it's a manual command on the CLI, I suggest we don't worry about that pre-mature caching that was done.

Reply:
And if we ever do get concerned about the performance of python string concatenation, [this might come in handy](https://gist.github.com/holmanb/84be00eab35477565cb95a1d62a741a9).